### PR TITLE
Fixing git style subcommands on Windows & allowing subcommands to come from dependency modules

### PR DESCRIPTION
**Pull request by mateodelnorte**
This PR fixes git style subcommands to work correctly on Windows, and adds logic such that git style subcommands can be inherited from child dependencies, allowing you to break large sets of commands into a multi-repo project. (It should also allow for dynamically varying available subcommands according to what is currently installed under ./node_modules)

I built a command for managing multi-repo projects, which dogfoods this patch to manage its own repos as well: https://github.com/mateodelnorte/meta/blob/master/package.json#L38-L39. meta, itself, has only one command – `meta`, which loads all child commands by searching through it's node_modules for matching `meta-*` modules and registering them via their index.js. 

This all works fine with the current `commander.js`, but when attempting to call them `commander` can't locate them. This patch ensures that `commander` can find the git subcommands and execute them. 
```
meta --help

  Usage: meta [options] [command]


  Commands:

    git         manage your meta repo and child git repositories 
    npm         run npm commands against your meta and child repositories 
    help [cmd]  display help for [cmd]

  Options:

    -h, --help     output usage information
    -V, --version  output the version number
```
The `git` subcommand, above, is provided by https://github.com/mateodelnorte/meta-git. 
The `npm` subcommand, above, is provided by https://github.com/mateodelnorte/meta-npm

### CODE REVIEW

**Comment by mateodelnorte** [comment]
Do we actually care that this breaks old (v<=0.0.12) versions of node?

**Comment by mateodelnorte** [comment]
PING PING

Are Node 0.8 and 0.6 support really required in commander.js? The only thing keeping this PR from passing all tests are engine restrictions in some dependencies. 

Thanks.

**Comment by mateodelnorte** [comment]
Hi. Howdy. Hello there. This is a pretty popular package. Lots of PR's. Some are months old. Do people respond to them?

**Comment by abetomo** [comment]
@mateodelnorte Would you please resolve the conflict?

**Comment by mateodelnorte** [comment]
Conflicts resolved.

File: index.js [hunk 1]
```diff
@@ -546,8 +549,15 @@ Command.prototype.executeSubCommand = function(argv, args, unknown) {
...
     }
   } else {
-    args.unshift(bin);
-    proc = spawn(process.execPath, args, { stdio: 'inherit'});
+    if (isExplicitJS) {
+      args.unshift(bin);
+      // add executable arguments to spawn
+      args = (process.execArgv || []).concat(args);
+
+      proc = spawn('node', args, { stdio: 'inherit'});
```
Comment (abetomo) [inline index.js]: As for 'node', can you do it like https: //github.com/tj/commander.js/blob/master/index.js#L544?

**Comment by abetomo** [comment]
Thank you!

File: index.js (hunk 1 above)
Comment (mateodelnorte) [inline index.js]: Should be good to go! Take a peek.

**Comment by mateodelnorte** [comment]
Glad to help! 

*not sure what's up with npm 5 not installing the single find-module-bin dependency. yarn installs it just find.

**Comment by mateodelnorte** [comment]
Any timeline on when this will be published to npm? I think a simple rebuild will fix the build errors on node 8. Seemed to be a temporary problem with the registry. The dependency version def exists: 
```
{ name: 'find-module-bin',
  description: 'scan known node_modules directories to find installation directory of a particular bin file',
  'dist-tags': { latest: '0.0.2' },
  versions: [ '0.0.0', '0.0.1', '0.0.2' ],
```

**Comment by mateodelnorte** [comment]
When is this going to be merged? The error happening in this build was a bug in npm5.

**Comment by mateodelnorte** [comment]
I've fixed the merge issue that happened in package.json as a result of this not being merged since January. Looks like a re-run of the builds works completely (because it was a bug in npm5). 

Can we get this merged?

Thx

**Comment by mateodelnorte** [comment]
@abetomo Can we please slate this for merge? It's been well over a year.

**Comment by abetomo** [comment]
@mateodelnorte 
It was awaiting @vanesyan's approval.

@vanesyan 
Could you please review it?

**Comment by mateodelnorte** [comment]
So... I updated this from upstream again. @abetomo @vanesyan 

Can we get this merged?

**Comment by mateodelnorte** [comment]
ping ping

File: index.js [hunk 2]
```diff
@@ -544,26 +545,23 @@ Command.prototype.executeSubCommand = function(argv, args, unknown) {
   if (exists(localBin + '.js')) {
     bin = localBin + '.js';
     isExplicitJS = true;
-  } else if (exists(localBin)) {
+  } else if (process.platform === 'win32' ? exists(localBin + '.cmd') : exists(localBin)) {
```
Comment (vanesyan) [inline index.js]: Can you please put it to own variable, for readability purpose

File: index.js [hunk 3]
```diff
@@ -5,6 +5,7 @@
 var EventEmitter = require('events').EventEmitter;
 var spawn = require('child_process').spawn;
 var path = require('path');
+var findModuleBin = require('find-module-bin');
```
Comment (vanesyan) [inline index.js]: Can you please inline the module, we trying to keep commander.js dependencies-free

**Comment by vanesyan** [comment]
Hello there! Sorry for make you waiting. Can you, please, resolve the comments in the review, thanks!

File: index.js (hunk 2 above)
Comment (mateodelnorte) [reply index.js]: Sure thing.

File: index.js (hunk 3 above)
Comment (mateodelnorte) [reply index.js]: As a curiosity – What's the reason for wanting no dependencies?

Comment (mateodelnorte) [reply index.js]: find-module-bin uses global-modules, to look up the appropriate os-dependent location of globally installed node modules. this won't be a simple pull-function-up refactor, as the os-specific logic is hidden in here. 

perhaps we can merge this and set the refactoring as tech-debt?

personally, I think the reuse is better for the community. you don't want a massive explosion of dependencies, but having well managed dependencies can make your code more DRY and keeps the potential for the dependent packages to continue to evolve and solidify.

Comment (mateodelnorte) [reply index.js]: @vanesyan

Comment (mateodelnorte) [reply index.js]: @vanesyan I don't think it makes sense to inline all of this. 

This PR has been ready for merge for about a year and six months.

**Comment by mateodelnorte** [comment]
Git style subcommands are still broken on Windows in master. This PR fixes them. 

Is there any plan to merge? Is the fact this is not merged because it's more important to have zero dependencies than to fix a glaring bug?

File: index.js (hunk 3 above)
Comment (codyaverett) [reply index.js]: Hi @vanesyan @abetomo
My team is also interested in this fix.  

Is there any reason why this can not be included with the dependency?

**Comment by shadowspawn** [comment]
@codyaverett What aspect of this Pull Request is your team interested in?

**Comment by shadowspawn** [comment]
@mateodelnorte Sorry this has been open for so long, but unfortunately I think I see why. There are some good and interesting ideas here (thanks!), but multiple potential problems from the point of view of a cautious maintainer. To be clear, I am not willing to recommend this Pull Request be accepted in its current form based on my current understanding.

Some initial comments and questions, but I warn you and apologise in advance there will be more! If you don't want to invest in engaging with a new maintainer that is ok, you proposed this a long time ago. I see ideas in here I want to consider further whether or not you abandon this Pull Request.

Side notes: 
- I suggest you don't put effort into resolving the conflicts yet unless you want to.
- I would like to add a CONTRIBUTING guide, and some of this discussion might end up there.

This is more a comment than a question or call for action. If you are addressing an issue or adding an enhancement that other people are asking for, it helps to link to the related issues. Otherwise the maintainers are judging the merit of the Pull Request just on your comments and what we recall of the issues we have seen in the past.

1) Are you doing one thing in this Pull Request or two? Do you consider the Windows "fix" to be separate, or is it an additional lookup method for binaries like the node_modules search?

2) In what way are Git style subcommands still broken on Windows? Are you referring to `*.cmd` or a wider issue?

**Comment by shadowspawn** [comment]
>Is there any reason why this can not be included with the dependency?

This is going to come up again with Pull Requests. My favourite two challenging images:

- https://devrant.com/rants/1854993/package-tsunami
- https://dev.to/leoat12/the-nodemodules-problem-29dc 

Commander currently has zero dependencies. That isn't  a requirement, but is a simple story. It does mean every suggested dependency will be critically evaluated.

And some insight into what this dependency pulls in:
- https://bundlephobia.com/result?p=commander@2.20.0 
- https://bundlephobia.com/result?p=find-module-bin@1.0.0

**Comment by mateodelnorte** [comment]
This PR was made LONG before Commander had zero dependencies. 

Yes, this PR 1) fixed git style subcommands on Windows and 2) introduced the ability for dependencies to supply git style subcommands. However, this was all done two years before the CONTRIBUTING guide you're proposing and I put in the effort to make PR merge-able multiple times. 

All well and good if you don't want to merge, but this entire process has been extremely disheartening.

**Comment by mateodelnorte** [comment]
Frankly, I don't care if this code gets merged. But I would love the two issues be fixed.

**Comment by codyaverett** [comment]
@shadowspawn Thank you for your time and detailed response.

My motive is directly correlated with this other old issue. https://github.com/mateodelnorte/meta/issues/46

I believe I can work around it, but I wasn't successful in my first attempts.  I  am still hopeful though. 😅 

I will report future related issues in the meta project directly. Thanks!

**Comment by shadowspawn** [comment]
Thanks for info @mateodelnorte and @codyaverett 

Sorry you had a rough time with this one @mateodelnorte.

I'm closing this Pull Request as problematic, but have noted it in #944 for future reference.
 
(I think the next direction we want to explore with git-style commands is more control over how the commands are found. Source file naming and directory layout is one form of that, but search paths where the commands are looked for is another as in this Pull Request. i.e. installed node_modules, search path for commands like npm uses, etc.)

**Pull request closed without merging**